File: 2874-maximum-value-of-an-ordered-triplet-ii/2874-maximum-value-of-an-ordered-triplet-ii.java
class Solution {
    public long maximumTripletValue(int[] nums) {
        int n=nums.length;
        int ans=0;
        int[] right_max = new int[n];
        int rm=nums[n-1];
        for(int i=n-2;i>=1;i--)
        {
            right_max[i]=rm;
            rm=Math.max(rm,nums[i]);
        }
        int left_max=nums[0];
        for(int i=1;i<n-1;i++)
        {
            ans=Math.max(ans,(left_max-nums[i])*right_max[i]);
            left_max=Math.max(left_max,nums[i]);
        }
        return ans;
    }
}

File: 2874-maximum-value-of-an-ordered-triplet-ii/2874-maximum-value-of-an-ordered-triplet-ii.py
class Solution:
    def maximumTripletValue(self, nums: List[int]) -> int:
        n=len(nums)
        max_triplet=0
        max_difference=0
        max_left=0
        for i in range(n):
            max_triplet=max(max_triplet,max_difference*nums[i])
            max_difference=max(max_difference,max_left-nums[i])
            max_left=max(max_left,nums[i])
        return max_triplet
            

        